app/jobs/scheduled/reindex_search.rb

```diff
@@ -1,10 +1,69 @@
+require_dependency 'search/query_limiter'
+
 module Jobs
   # if locale changes or search algorithm changes we may want to reindex stuff
   class ReindexSearch < Jobs::Scheduled
     every 1.day
 
     def execute(args)
-      Search.rebuild_problem_posts
+      limiter = Search::QueryLimiter.new(args[:limit] || 10_000)
+      limiter.step { |limit| rebuild_problem_categories(limit) }
```

Comment:
I don't really think we need all this fancy...  just limit to 10000 topics / 10000 posts / 10000 users. no need to do a limit on 10000 things.  code will be much easier to understand, one less class.